mesa/mesaPROTON_OC.py

```diff
@@ -569,26 +570,40 @@ def pick_from_population_pool_by_age(self, age_wanted):
         self.population.remove(picked_person)
         return picked_person
 
-    def df_to_dict(self, df):
+    def df_to_dict(self, df, extra_depth=False):
         """
-        Based on the size of the dataframe, it transforms the dataframe into a nested dictionary.
-        :param df: pandas dataframe
-        :return: dic
+        Based on the number of pandas DataFrame columns, transforms the dataframe into nested dictionaries as follows:
+        df-columns = age, sex, education, p --> dict-keys = {age:{sex:[education, p]}}
+
+        If extra_depth is True the transformation has an extra level of depth as follows:
+        df-columns = age, sex, education, p --> dict-keys = {age:{sex:{education: p}}}
+
+        This transformation ensures a faster access to the values using the dictionary keys.
+        :param df: pandas df, the df to be transformed
+        :param extra_depth: bool, if True gives an extra level of depth
+        :return: dict, a new dictionary
         """
         dic = dict()
-        if len(df.columns) == 2:
+        extra_depth_modifier = 0
+        if extra_depth:
+            extra_depth_modifier = 1
+
+        if len(df.columns) + extra_depth_modifier == 2:
             for col in np.unique(df.iloc[:,0]):
-                dic[col] = float(df[df.iloc[:,0] == col].iloc[:,1].values)
-        if len(df.columns) == 3:
+                dic[col] = df[df.iloc[:,0] == col].iloc[:,1].values
+        if len(df.columns) + extra_depth_modifier == 3:
             for col in np.unique(df.iloc[:,0]):
                 dic[col] = df[df.iloc[:, 0] == col].iloc[:, 1:].values
-        if len(df.columns) == 4:
+        if len(df.columns) + extra_depth_modifier == 4:
             for col in np.unique(df.iloc[:, 0]):
                 dic[col] = df[df.iloc[:, 0] == col].iloc[:, 1:]
             for key in dic:
                 subdic = dict()
                 for subcol in np.unique(dic[key].iloc[:, 0]):
-                    subdic[subcol] = dic[key][dic[key].iloc[:, 0] == subcol].iloc[:, 1:].values
+                    if extra_depth:
```

Comment:
Can you try to avoid repeated code? Maybe the part `dic[col] = df[df.iloc[:,0] == col].iloc[:,1].values` can stay out of the if, while the subdict appears only `if extra_depth and len(df.columns)>3` (note that this would remove the need for the `extra_depth_modifier` variable)

Reply:
what do you think of this new solution?